# Genomic assessment of invasion dynamics of SARS-CoV-2 Omicron BA.1

​​**Joseph L.-H. Tsui**, John T. McCrone, Ben Lambert, Sumali Bajaj, Rhys P.D. Inward, Paolo Bosetti, Houriiyah Tegally, Verity Hill, Rosario Evans Pena, Alexander E. Zarebski, Thomas P. Peacock, Luyang Liu, Neo Wu, Megan Davis, Isaac I. Bogoch, Kamran Khan, Meaghan Kall, Nurin Iwani Binti Abdul Aziz, Rachel Colquhoun, Áine O’Toole, Ben Jackson, Abhishek Dasgupta, Eduan Wilkinson, Tulio de Oliveira, The COVID-19 Genomics UK (COG-UK) consortium, Thomas R. Connor, Nicholas J. Loman, Vittoria Colizza, Christophe Fraser, Erik Volz, Xiang Ji, Bernardo Gutierrez, Meera Chand, Simon Dellicour, **Simon Cauchemez**, **Jayna Raghwani**, **Marc A. Suchard**, **Philippe Lemey**, **Andrew Rambaut**, **Oliver G. Pybus**, **Moritz U.G. Kraemer**

[![DOI](https://zenodo.org/badge/638929927.svg)](https://zenodo.org/badge/latestdoi/638929927)

---

This repository contains data and scripts used to generate results
presented in [citation]. Please note that some data files cannot be shared publicly due to data sharing agreements; some scripts may also need adjustments to run depending on local setup.

GISAID [terms of use](https://www.gisaid.org/registration/terms-of-use/) prohibit sharing of sequence data in this repository. Therefore we make the GISAID accessions available here and provide a table of acknowledgements. Note also that we cannot make administrative level two (adm2) metadata for genomic sequences from COG-UK (https://www.cogconsortium.uk) available.

## Abstract

_SARS-CoV-2 variants of concern (VOCs) now arise in the context of heterogeneous human connectivity and population immunity. Through a large-scale phylodynamic analysis of 115,622 Omicron BA.1 genomes, we identified >6,000 introductions of the antigenically-distinct VOC into England and analysed their local transmission and dispersal history. We find that six of the eight largest English Omicron lineages were already transmitting when Omicron was first reported in southern Africa (22 November 2021). Multiple data sets show importation of Omicron continued despite subsequent restrictions on travel from southern Africa, due to export from well-connected secondary locations. Initiation and dispersal of Omicron transmission lineages in England was a two-stage process that can be explained by models of the country’s human geography and hierarchical travel network. Our results enable a comparison of the processes that drive the invasion of Omicron and other VOCs across multiple spatial scales._

## Repository usage and structure

The structure of this repository is shown below:

```
├── analyses
│   ├── epidemiological
│   ├── importation
│   ├── continuous-phylogeography
│   ├── discrete-phylogeography-GLM
│   └── README.md
├── data
│   ├── phylogenetic
│   ├── epidemiological
│   ├── spatial-geographic
│   └── README.md
├── LICENSE.md
└── README.md
```

### Input data

All input data we are able to share publicly are stored in the [`data`](data/) directory.

### Analyses

The [`analyses`](analyses/) directory contains sub-directories for the following analyses:

- **Epidemiological analyses:** [`epidemiological`](analyses/epidemiological/) contains scripts for processing and visualising data relevant to the epidemiology of and human mobility patterns during the Omicron BA.1 wave.
- **Importation analyses:** [`importation`](analyses/importation/) contains scripts and other materials for analyses relevant to Omicron BA.1 importation dynamics, including the phylodynamic analysis, the branching process model, and also calculating the Estimated Importation Intensity (EII) for potential exporters.
- **Continous phylogeographic analyses:** [`continuous-phylogeography`](analyses/continuous-phylogeography/) contains scripts and XMLs for continuous phylogeographic reconstruction of identified transmission lineages within England; also contains scripts for visualising results from these analyses.
- **Discrete phylogeographic analyses with GLM:** [`discrete-phylogeography-GLM`](analyses/discrete-phylogeography-GLM/) contains scripts and XMLs for discrete phylogeographic analyses with GLM to test the association of demographic, epidemiological, and mobility-related factors with viral lineage movements at the LTLA-level; also contains scripts for visualising results from these analyses.

## Data Availability

The Google COVID-19 Aggregated Mobility Research Dataset is not publicly available owing to stringent licensing agreements. Information on the process of requesting access to the Google mobility data is available from aaronschneider@google.com.

UK genome sequences used were generated by the COG-UK consortium (https://www.cogconsortium.uk/). Data linking COG-IDs to location have been removed to protect privacy, however if you require this data, please visit https://www.cogconsortium.uk/contact/ for information on accessing consortium-only data.

---
